### PR TITLE
Integration test framework

### PR DESCRIPTION
**Pull request by ethantkoenig**
__NOTE__: Drone is currently failing due to `.drone.yml.sig`, which needs to be updated.

A new framework for integration tests. The main issues with our current integration tests are that it
- Uses the production database
- Re-installs the server for every test
- Must manually add content (e.g. users, repos) to the system for every test

My proposal is to have a test "environment" (database and filesystem containing repos). The integration test environment would be populated with "fixtures", somewhat similar to what we do for unit tests (plus actual repos in the filesystem). Right now, this is implemented by having a separate `app.ini` for integration tests, as well as a collection of database fixtures and directory of repositories.

Also implements @andreynering's suggestion from https://github.com/go-gitea/gitea/pull/1135#discussion_r105298393 of using the `ServeHTTP(..)` method

This system would allow us to not only add integration tests, but also unit tests for components that interact with the filesystem (e.g. anything that runs git commands).

__STORING THE INTEGRATION ENVIRONMENT__:
In order to not have to store a bunch of test fixtures and repos inside the main gitea repository, I have created a separate repo (https://github.com/ethantkoenig/gitea-integration) that stores the integration test fixtures and repositories. This repo must be cloned to run integration tests. If we end up choosing this approach, it might make sense to transfer this repo to the `go-gitea` organization.

__WHAT TO DO WITH EXISTING INTEGRATION TESTS__:
For now, I have removed the existing integration tests to clear up the `integrations/` directory. I have ported all previous integration tests to the new framework.

__HOW TO RUN INTEGRATION TESTS__:
1. `make test-(sqlite|mysql|pgsql)`. You may need to update the corresponding `integrations/*.ini` to match whatever you have running locally.

### CODE REVIEW

File: cmd/web.go [hunk 1]
```diff
@@ -71,7 +71,7 @@ and it takes care of all the other things for you`,
 }
 
 // newMacaron initializes Macaron instance.
-func newMacaron() *macaron.Macaron {
+func NewMacaron() *macaron.Macaron {
```
Comment (bkcsoft) [inline cmd/web.go]: Can we move this (and `RegisterRoutes` to a new `modules/router` instead? Since they really shouldn't be here anyway 🙂 )

File: models/models.go [hunk 2]
```diff
@@ -24,6 +24,9 @@ import (
 	// Needed for the MSSSQL driver
 	_ "github.com/denisenkom/go-mssqldb"
 
+	// Needed for SQLite
+	_ "github.com/mattn/go-sqlite3"
```
Comment (bkcsoft) [inline models/models.go]: nononononono, there's a reason that this _isn't_ here...

File: myintegrations/gitea-repositories/user1/repo1.git/hooks/post-receive.d/gitea [hunk 3]
```diff
@@ -0,0 +1,2 @@
+#!/usr/bin/env bash
+"/home/ethantkoenig/.gopath/src/code.gitea.io/gitea/gitea" hook --config='integrations/app.ini' post-receive
```
Comment (bkcsoft) [inline myintegrations/gitea-repositories/user1/repo1.git/hooks/post-receive.d/gitea]: This will break for everyone... Don't put the test-repos in here, have the test pull it from somewhere (or untar/unzip)

**Review by bkcsoft — CHANGES_REQUESTED** [review]

**Comment by bkcsoft** [comment]
Also don't call it `myintegrations`, just go for `intergrations` or `tests/integration` :)

File: models/models.go (hunk 2 above)
Comment (ethantkoenig) [reply models/models.go]: I know, I added it as a temporary hack so the tests would work; I realize that sqlite has to be explicitly enabled. Perhaps I should have made that clearer.

File: myintegrations/gitea-repositories/user1/repo1.git/hooks/post-receive.d/gitea (hunk 3 above)
Comment (ethantkoenig) [reply myintegrations/gitea-repositories/user1/repo1.git/hooks/post-receive.d/gitea]: Hmm, I didn't realize there were hard-coded absolute paths in here. If we pull from somewhere else, we'll still have to edit all of these scripts to have the correct path (which is doable, but annoying).

File: cmd/web.go (hunk 1 above)
Comment (ethantkoenig) [reply cmd/web.go]: Done

File: models/models.go (hunk 2 above)
Comment (ethantkoenig) [reply models/models.go]: I figured out the non-hacky way to make things work, so this has been removed.

**Comment by ethantkoenig** [comment]
Rebased, renamed `myintegrations` to `integrations` and remove the previous contents of `integrations`.

**Comment by lunny** [comment]
It seems #741 had imported the integration tests and #1135 added signup tests.

**Comment by strk** [comment]
I'd like the integration test framework to also let me test PostgreSQL
as a backend (for example). So I'd prefer fixtures to be passed to the
ORM to fill the target DB of choice for testing.

**Comment by ethantkoenig** [comment]
@lunny I am aware, but there are some problems with these integration tests (see PR description). It was my understanding that #741 and #1135 were not a permanent solution (https://github.com/go-gitea/gitea/pull/1135#issuecomment-285589648).

**Comment by ethantkoenig** [comment]
Rebased to resolve conflicts. 

@strk I've updated it to use testfixtures, so to use another RDBMS all you have to do is modify `integrations/app.ini`

File: myintegrations/gitea-repositories/user1/repo1.git/hooks/post-receive.d/gitea (hunk 3 above)
Comment (ethantkoenig) [reply myintegrations/gitea-repositories/user1/repo1.git/hooks/post-receive.d/gitea]: I've updated the scripts to use `$GITEA_ROOT` instead of a hard-coded path. We'll have to export the appropriate value for `GITEA_ROOT` before running integration tests involving these scripts.

**Comment by strk** [comment]
@ethantkoenig would it be possible to script the app.ini update so to have `make check` (or `go test` if possible) run the test against all supported (at build-time) database backends ?

**Comment by ethantkoenig** [comment]
@strk I've created separate `.ini` files for SQLite, MySQL, and PostgreSQL. The integration tests can be run using `make test-(sqlite|mysql|pgsql)`. All three pass locally, but I've been having some trouble getting MySQL and PostgreSQL to pass in CI.

File: .gitignore [hunk 4]
```diff
@@ -44,3 +44,5 @@ coverage.out
 /indexers
 /log
 /public/img/avatar
+
```
Comment (strk) [inline .gitignore]: empty line ?

File: .drone.yml [hunk 5]
```diff
@@ -42,7 +53,7 @@ pipeline:
       TAGS: bindata
       GOPATH: /srv/app
     commands:
-      - make test-pgsql
+      - echo 'Not ready yet'
```
Comment (strk) [inline .drone.yml]: Maybe you could leave the intended command in here, for reference, like `echo make 'test-pgsql #not ready yet'` (and same for all non-ready blocks)

File: .drone.yml [hunk 6]
```diff
@@ -24,14 +24,25 @@ pipeline:
...
       event: [ push, tag, pull_request ]
 
+  test-sqlite:
+    image: webhippie/golang:edge
+    pull: true
+    environment:
+      TAGS: bindata
+      GOPATH: /srv/app
+    commands:
+      - echo 'Not ready yet'
```
Comment (strk) [inline .drone.yml]: I thought you mentioned this was actually passing on drone ?

**Review by strk — CHANGES_REQUESTED** [review]
Please don't add that tar.gz to the repository (what is it for?).

Also, I don't understand the changes to cmd/web.go

**Comment by ethantkoenig** [comment]
The `.tar.gz` was accidentally added; I've removed it, and added it to `.gitignore`.

The changes to `cmd/web.go` are following @bkcsoft's suggestion above

**Comment by andreynering** [comment]
Drone error seems unrelated to the change:

> WARNING: unable to verify the Yaml signature.

**Comment by andreynering** [comment]
LGTM :tada: 

I'd just disable Xorm log output on console.

File: .drone.yml (hunk 6 above)
Comment (ethantkoenig) [reply .drone.yml]: Fixed.

File: .gitignore (hunk 4 above)
Comment (ethantkoenig) [reply .gitignore]: Removed.

**Comment by ethantkoenig** [comment]
@andreynering Do you know how to update `.drone.yml.sig` to match `.drone.yml`?

**Review by bkcsoft — APPROVED** [review]

**Comment by ethantkoenig** [comment]
Does anyone know how to generate a new `.drone.yml.sig`? When I try running `drone sign go-gitea/gitea`, I get `Error: you must provide the Drone server address.`

**Comment by sfnke** [comment]
@ethantkoenig you have to export environment variables:

```
export DRONE_SERVER="https://drone.example.com"
export DRONE_TOKEN="myDroneToken"
```

**Comment by appleboy** [comment]
@ethantkoenig @tboerger We need to remove `.drone.yml.sig`. Drone support secret interpolation in `0.5` version and remove `--image` flag on secret command.

**Comment by bkcsoft** [comment]
@appleboy signature is still required for secrets 😒

**Comment by ethantkoenig** [comment]
Rebased to resolve conflicts.

Assuming my changes to `.drone.yml` are acceptable, could some with the drone token please update `.drone.yml.sig`?

**Comment by appleboy** [comment]
@ethantkoenig @bkcsoft Only @tboerger has permission to update `.drone.yml.sig`. BTW drone drop the `.drone.yml.sig` file on 0.6 version.

**Comment by ethantkoenig** [comment]
Rebased to resolve conflict in `cmd/web.go`.

Would it be worth not updating `.drone.yml` in this PR to help move things along? It's currently blocked on getting `.drone.yml.sig` updated.

**Comment by lunny** [comment]
@tboerger maybe help you about `.drone.yml.sig`

**Comment by strk** [comment]
As @tboerger declared on Gitter he's stepping out of Ownership role, we need others to step up to deal with signature, or drop the required signature altogether (I'm not sure what it is needed for)

**Comment by strk** [comment]
Why is `modules/router/router.go` used rather than something under `routers/` ?

**Comment by ethantkoenig** [comment]
@bkcsoft's suggestion, I could move it elsewhere if you want.

**Comment by lunny** [comment]
conflicts

**Comment by strk** [comment]
And drone fails with:
```
modules/router/router.go:620: undefined: private in private.RegisterRoutes
```

**Comment by ethantkoenig** [comment]
Rebased to resolve conflict, and fixed missing import bug.

**Comment by strk** [comment]
@bkcsoft why `modules/router/router.go` rather than `routers/something` ?

File: Makefile [hunk 7]
```diff
@@ -79,13 +79,27 @@ integrations: build
...
-test-pgsql:
-	@echo "Not integrated yet!"
+test-pgsql: integrations.test integrations/gitea-integration
+	GITEA_CONF=integrations/pgsql.ini ./integrations.test
+
+integrations.test: $(SOURCES)
+	go test -c code.gitea.io/gitea/integrations -tags 'sqlite'
+
+integrations/gitea-integration:
+	curl -L https://github.com/ethantkoenig/gitea-integration/archive/v2.tar.gz > integrations/gitea-integration.tar.gz
```
Comment (lunny) [inline Makefile:106]: What's the archive?

File: integrations/integration_test.go [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+	"net/http"
+	"os"
+	"testing"
+
+	"code.gitea.io/gitea/models"
+	"code.gitea.io/gitea/modules/router"
+	"code.gitea.io/gitea/modules/setting"
+	"code.gitea.io/gitea/routers"
+
+	"bytes"
```
Comment (lunny) [inline integrations/integration_test.go]: wrong import place

File: integrations/signup_test.go [hunk 9]
```diff
@@ -2,34 +2,39 @@
...
-	"os"
+	"bytes"
+	"net/http"
+	"net/url"
 	"testing"
 
-	"code.gitea.io/gitea/integrations/internal/utils"
+	"code.gitea.io/gitea/models"
+	"code.gitea.io/gitea/modules/setting"
+	"github.com/stretchr/testify/assert"
```
Comment (lunny) [inline integrations/signup_test.go]: blank line.

File: integrations/view_test.go [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+// license that can be found in the LICENSE file.
+
+package integrations
+
+import (
+	"net/http"
+	"testing"
+
+	"code.gitea.io/gitea/models"
+	"github.com/stretchr/testify/assert"
```
Comment (lunny) [inline integrations/view_test.go]: blank line

**Review by lunny — CHANGES_REQUESTED** [review]

**Comment by bkcsoft** [comment]
@strk @ethantkoenig I didn't really have a reason for `modules/routes/` over `routers/`, feel free to move it. But it didn't belong as an exported function in `cmd/web.go` (_nothing_ should import `/cmd` except for `main.go` :wink: )

File: Makefile (hunk 7 above)
Comment (ethantkoenig) [reply Makefile:106]: It contains the git repositories for the integration tests. It is necessary, because some integration may need to run git commands, or access files in the repositories.

See also the "STORING THE INTEGRATION ENVIRONMENT" section of the PR description.

**Comment by ethantkoenig** [comment]
@lunny I've addressed your review, and moved `modules/router/router.go` to `routers`

**Review by lunny — APPROVED** [review]

**Comment by lunny** [comment]
LGTM

File: .drone.yml [hunk 11]
```diff
@@ -24,14 +24,25 @@ pipeline:
...
+
   test-mysql:
     image: webhippie/golang:edge
     pull: true
     environment:
       TAGS: bindata
       GOPATH: /srv/app
     commands:
-      - make test-mysql
+      - echo make test-mysql # Not ready yet
```
Comment (appleboy) [inline .drone.yml:46]: revert this changes? @ethantkoenig

File: .drone.yml [hunk 12]
```diff
@@ -42,7 +53,7 @@ pipeline:
       TAGS: bindata
       GOPATH: /srv/app
     commands:
-      - make test-pgsql
+      - echo make test-pqsql # Not ready yet
```
Comment (appleboy) [inline .drone.yml:57]: revert this changes? @ethantkoenig

File: .drone.yml (hunk 11 above)
Comment (ethantkoenig) [reply .drone.yml:46]: `test-mysql` and `test-pgsql` still aren't ready to run in CI (although they do pass locally). Unfortunately, I've had a hard time debugging why they aren't passing in CI, and given how long it took to get a `.drone.yml.sig` for this PR, I think makes more sense to address them in a future PR.

Comment (ethantkoenig) [reply .drone.yml:46]: See also https://github.com/go-gitea/gitea/pull/1290#discussion_r107802335

**Comment by ethantkoenig** [comment]
@strk I believe I have addressed all of your requested changes.

File: .drone.yml (hunk 11 above)
Comment (appleboy) [reply .drone.yml:46]: OK

**Comment by strk** [comment]
Thanks, all checks pass now, just needs to be merged. Looking forward to enable the CI testing of mysql and postgresql !

**Comment by lunny** [comment]
conflicted. @ethantkoenig

**Comment by ethantkoenig** [comment]
@lunny Rebased to resolve conflicts. Unfortunately, I will need a new `.drone.yml.sig` now that #1468 has been merged into master.

**Comment by lunny** [comment]
@ethantkoenig I have resigned it.
